File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.java
class Solution {
    int moves = 0;
    
    public int distributeCoins(TreeNode root) {
        dfs(root);
        return moves;
    }

    // return -ve if node is accepting and +ve if it is giving back
    public int dfs(TreeNode root) {
        if(root == null) return 0;

        var leftBalance = dfs(root.left);
        var rightBalance = dfs(root.right);

        var curr = leftBalance + rightBalance + root.val - 1;
        moves += (Math.abs(curr));
        return curr;
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.go
var moves int

func distributeCoins(root *TreeNode) int {
    moves = 0
    dfs(root)
    return moves
}

// return -ve if node is accepting and +ve if it is giving back
func dfs(root *TreeNode) int {
    if root == nil {
        return 0
    }

    leftBalance := dfs(root.Left)
    rightBalance := dfs(root.Right)

    curr := leftBalance + rightBalance + root.Val - 1
    moves += int(math.Abs(float64(curr)))
    return curr
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */